## Human Activity Recognition with Smartphones
Recordings of 30 study participants performing activities of daily living

### About Dataset
The Human Activity Recognition database was built from the recordings of 30 study participants performing activities of daily living (ADL) while carrying a waist-mounted smartphone with embedded inertial sensors. The objective is to classify activities into one of the six activities performed.

### Description of experiment
The experiments have been carried out with a group of 30 volunteers within an age bracket of 19-48 years. Each person performed six activities (WALKING, WALKING_UPSTAIRS, WALKING_DOWNSTAIRS, SITTING, STANDING, LAYING) wearing a smartphone (Samsung Galaxy S II) on the waist. Using its embedded accelerometer and gyroscope, we captured 3-axial linear acceleration and 3-axial angular velocity at a constant rate of 50Hz. The experiments have been video-recorded to label the data manually. The obtained dataset has been randomly partitioned into two sets, where 70% of the volunteers was selected for generating the training data and 30% the test data.

The sensor signals (accelerometer and gyroscope) were pre-processed by applying noise filters and then sampled in fixed-width sliding windows of 2.56 sec and 50% overlap (128 readings/window). The sensor acceleration signal, which has gravitational and body motion components, was separated using a Butterworth low-pass filter into body acceleration and gravity. The gravitational force is assumed to have only low frequency components, therefore a filter with 0.3 Hz cutoff frequency was used. From each window, a vector of features was obtained by calculating variables from the time and frequency domain.

In [404]:
import numpy as np
import pandas as pd

In [405]:
train = pd.read_csv(r'C:\Users\rishi\OneDrive\Desktop\Sabudh courseworks\FEATURE_SELECTION\kagglehub\datasets\uciml\human-activity-recognition-with-smartphones\versions\2\train.csv')
train.head(20)

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0            0.288585          -0.020294          -0.132905         -0.995279   
1            0.278419          -0.016411          -0.123520         -0.998245   
2            0.279653          -0.019467          -0.113462         -0.995380   
3            0.279174          -0.026201          -0.123283         -0.996091   
4            0.276629          -0.016570          -0.115362         -0.998139   
5            0.277199          -0.010098          -0.105137         -0.997335   
6            0.279454          -0.019641          -0.110022         -0.996921   
7            0.277432          -0.030488          -0.125360         -0.996559   
8            0.277293          -0.021751          -0.120751         -0.997328   
9            0.280586          -0.009960          -0.106065         -0.994803   
10           0.276880          -0.012722          -0.103438         -0.994815   
11           0.276228          -0.021441          -0.108202         -0.998246   
12           0.278457          -0.020415          -0.112732         -0.999135   
13           0.277175          -0.014713          -0.106756         -0.999188   
14           0.297946           0.027094          -0.061668         -0.988641   
15           0.279203          -0.023020          -0.122080         -0.996839   
16           0.279038          -0.014800          -0.116849         -0.996941   
17           0.280135          -0.013917          -0.106370         -0.997695   
18           0.277731          -0.018211          -0.109188         -0.997491   
19           0.275568          -0.016980          -0.111429         -0.997811   

    tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0          -0.983111         -0.913526         -0.995112         -0.983185   
1          -0.975300         -0.960322         -0.998807         -0.974914   
2          -0.967187         -0.978944         -0.996520         -0.963668   
3          -0.983403         -0.990675         -0.997099         -0.982750   
4          -0.980817         -0.990482         -0.998321         -0.979672   
5          -0.990487         -0.995420         -0.997627         -0.990218   
6          -0.967186         -0.983118         -0.997003         -0.966097   
7          -0.966728         -0.981585         -0.996485         -0.966313   
8          -0.961245         -0.983672         -0.997596         -0.957236   
9          -0.972758         -0.986244         -0.995405         -0.973663   
10         -0.973077         -0.985357         -0.995509         -0.973948   
11         -0.987214         -0.992727         -0.998251         -0.985997   
12         -0.984680         -0.996274         -0.999077         -0.982937   
13         -0.990526         -0.993365         -0.999211         -0.990687   
14         -0.816699         -0.901907         -0.988958         -0.794280   
15         -0.974848         -0.983386         -0.997094         -0.973332   
16         -0.981866         -0.982577         -0.997220         -0.981620   
17         -0.987516         -0.990407         -0.998014         -0.987954   
18         -0.993222         -0.996128         -0.997903         -0.992711   
19         -0.990522         -0.997621         -0.998205         -0.989470   

    tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
0          -0.923527         -0.934724  ...                        -0.710304   
1          -0.957686         -0.943068  ...                        -0.861499   
2          -0.977469         -0.938692  ...                        -0.760104   
3          -0.989302         -0.938692  ...                        -0.482845   
4          -0.990441         -0.942469  ...                        -0.699205   
5          -0.995549         -0.942469  ...                        -0.844619   
6          -0.983116         -0.940987  ...                        -0.564430   
7          -0.982982   

In [406]:
test = pd.read_csv(r'C:\Users\rishi\OneDrive\Desktop\Sabudh courseworks\FEATURE_SELECTION\kagglehub\datasets\uciml\human-activity-recognition-with-smartphones\versions\2\test.csv')
test.head(10)

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.257178          -0.023285          -0.014654         -0.938404   
1           0.286027          -0.013163          -0.119083         -0.975415   
2           0.275485          -0.026050          -0.118152         -0.993819   
3           0.270298          -0.032614          -0.117520         -0.994743   
4           0.274833          -0.027848          -0.129527         -0.993852   
5           0.279220          -0.018620          -0.113902         -0.994455   
6           0.279746          -0.018271          -0.104000         -0.995819   
7           0.274601          -0.025035          -0.116831         -0.995594   
8           0.272529          -0.020954          -0.114472         -0.996784   
9           0.275746          -0.010372          -0.099776         -0.998373   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.920091         -0.667683         -0.952501         -0.925249   
1         -0.967458         -0.944958         -0.986799         -0.968401   
2         -0.969926         -0.962748         -0.994403         -0.970735   
3         -0.973268         -0.967091         -0.995274         -0.974471   
4         -0.967445         -0.978295         -0.994111         -0.965953   
5         -0.970417         -0.965316         -0.994585         -0.969481   
6         -0.976354         -0.977725         -0.995996         -0.973665   
7         -0.982069         -0.985262         -0.995341         -0.981485   
8         -0.975906         -0.986597         -0.997029         -0.973735   
9         -0.986933         -0.991022         -0.998663         -0.987140   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
0         -0.674302         -0.894088  ...                        -0.705974   
1         -0.945823         -0.894088  ...                        -0.594944   
2         -0.963483         -0.939260  ...                        -0.640736   
3         -0.968897         -0.938610  ...                        -0.736124   
4         -0.977346         -0.938610  ...                        -0.846595   
5         -0.965897         -0.937856  ...                        -0.857565   
6         -0.979253         -0.937856  ...                        -0.391286   
7         -0.984610         -0.941263  ...                        -0.672358   
8         -0.985556         -0.941263  ...                        -0.853248   
9         -0.991084         -0.943761  ...                        -0.741435   

   angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                     0.006462                              0.162920   
1                    -0.083495                              0.017500   
2                    -0.034956                              0.202302   
3                    -0.017067                              0.154438   
4                    -0.002223                             -0.040046   
5                    -0.095681                              0.048849   
6                    -0.309083                             -0.000857   
7                    -0.025526                             -0.214033   
8                     0.144626                             -0.035564   
9                     0.157584                             -0.033487   

   angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
0                         -0.825886                              0.271151   
1                         -0.434375                              0.920593   
2                          0.064103                              0.145068   
3                          0.340134                              0.296407   
4                          0.736715                             -0.118545   
5                          0.760684                             -0.072216   
6                          0.728514                             -0.33377

In [407]:
train.columns

Index(['tBodyAcc-mean()-X', 'tBodyAcc-mean()-Y', 'tBodyAcc-mean()-Z',
       'tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X',
       ...
       'fBodyBodyGyroJerkMag-kurtosis()', 'angle(tBodyAccMean,gravity)',
       'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'subject', 'Activity'],
      dtype='object', length=563)

In [408]:
train.shape

(7352, 563)

In [409]:
test.shape

(2947, 563)

In [410]:
train.isnull().sum()

tBodyAcc-mean()-X       0
tBodyAcc-mean()-Y       0
tBodyAcc-mean()-Z       0
tBodyAcc-std()-X        0
tBodyAcc-std()-Y        0
                       ..
angle(X,gravityMean)    0
angle(Y,gravityMean)    0
angle(Z,gravityMean)    0
subject                 0
Activity                0
Length: 563, dtype: int64

In [411]:
x_train = train.iloc[:,:-1]
x_train


tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
0              0.288585          -0.020294          -0.132905   
1              0.278419          -0.016411          -0.123520   
2              0.279653          -0.019467          -0.113462   
3              0.279174          -0.026201          -0.123283   
4              0.276629          -0.016570          -0.115362   
...                 ...                ...                ...   
7347           0.299665          -0.057193          -0.181233   
7348           0.273853          -0.007749          -0.147468   
7349           0.273387          -0.017011          -0.045022   
7350           0.289654          -0.018843          -0.158281   
7351           0.351503          -0.012423          -0.203867   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
0            -0.995279         -0.983111         -0.913526         -0.995112   
1            -0.998245         -0.975300         -0.960322         -0.998807   
2            -0.995380         -0.967187         -0.978944         -0.996520   
3            -0.996091         -0.983403         -0.990675         -0.997099   
4            -0.998139         -0.980817         -0.990482         -0.998321   
...                ...               ...               ...               ...   
7347         -0.195387          0.039905          0.077078         -0.282301   
7348         -0.235309          0.004816          0.059280         -0.322552   
7349         -0.218218         -0.103822          0.274533         -0.304515   
7350         -0.219139         -0.111412          0.268893         -0.310487   
7351         -0.269270         -0.087212          0.177404         -0.377404   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
0            -0.983185         -0.923527         -0.934724  ...   
1            -0.974914         -0.957686         -0.943068  ...   
2            -0.963668         -0.977469         -0.938692  ...   
3            -0.982750         -0.989302         -0.938692  ...   
4            -0.979672         -0.990441         -0.942469  ...   
...                ...               ...               ...  ...   
7347          0.043616          0.060410          0.210795  ...   
7348         -0.029456          0.080585          0.117440  ...   
7349         -0.098913          0.332584          0.043999  ...   
7350         -0.068200          0.319473          0.101702  ...   
7351         -0.038678          0.229430          0.269013  ...   

      fBodyBodyGyroJerkMag-skewness()  fBodyBodyGyroJerkMag-kurtosis()  \
0                           -0.298676                        -0.710304   
1                           -0.595051                        -0.861499   
2                           -0.390748                        -0.760104   
3                           -0.117290                        -0.482845   
4                           -0.351471                        -0.699205   
...                               ...                              ...   
7347                        -0.588433                        -0.880324   
7348                        -0.390738                        -0.680744   
7349                         0.025145                        -0.304029   
7350                         0.063907                        -0.344314   
7351                        -0.387052                        -0.740738   

      angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                       -0.112754                              0.030400   
1                        0.053477                             -0.007435   
2                       -0.118559                              0.177899   
3                       -0.036788                             -0.012892   
4                        0.123320                              0.122542   
...                           ...                                   ...   
7347                    -0.190437            

In [412]:
y_train = train.iloc[:,-1]
y_train

0               STANDING
1               STANDING
2               STANDING
3               STANDING
4               STANDING
              ...       
7347    WALKING_UPSTAIRS
7348    WALKING_UPSTAIRS
7349    WALKING_UPSTAIRS
7350    WALKING_UPSTAIRS
7351    WALKING_UPSTAIRS
Name: Activity, Length: 7352, dtype: object

In [413]:
x_test = test.iloc[:,:-1]
x_test

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
0              0.257178          -0.023285          -0.014654   
1              0.286027          -0.013163          -0.119083   
2              0.275485          -0.026050          -0.118152   
3              0.270298          -0.032614          -0.117520   
4              0.274833          -0.027848          -0.129527   
...                 ...                ...                ...   
2942           0.310155          -0.053391          -0.099109   
2943           0.363385          -0.039214          -0.105915   
2944           0.349966           0.030077          -0.115788   
2945           0.237594           0.018467          -0.096499   
2946           0.153627          -0.018437          -0.137018   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
0            -0.938404         -0.920091         -0.667683         -0.952501   
1            -0.975415         -0.967458         -0.944958         -0.986799   
2            -0.993819         -0.969926         -0.962748         -0.994403   
3            -0.994743         -0.973268         -0.967091         -0.995274   
4            -0.993852         -0.967445         -0.978295         -0.994111   
...                ...               ...               ...               ...   
2942         -0.287866         -0.140589         -0.215088         -0.356083   
2943         -0.305388          0.028148         -0.196373         -0.373540   
2944         -0.329638         -0.042143         -0.250181         -0.388017   
2945         -0.323114         -0.229775         -0.207574         -0.392380   
2946         -0.330046         -0.195253         -0.164339         -0.430974   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
0            -0.925249         -0.674302         -0.894088  ...   
1            -0.968401         -0.945823         -0.894088  ...   
2            -0.970735         -0.963483         -0.939260  ...   
3            -0.974471         -0.968897         -0.938610  ...   
4            -0.965953         -0.977346         -0.938610  ...   
...                ...               ...               ...  ...   
2942         -0.148775         -0.232057          0.185361  ...   
2943         -0.030036         -0.270237          0.185361  ...   
2944         -0.133257         -0.347029          0.007471  ...   
2945         -0.279610         -0.289477          0.007471  ...   
2946         -0.218295         -0.229933         -0.111527  ...   

      fBodyBodyGyroJerkMag-skewness()  fBodyBodyGyroJerkMag-kurtosis()  \
0                           -0.330370                        -0.705974   
1                           -0.121845                        -0.594944   
2                           -0.190422                        -0.640736   
3                           -0.344418                        -0.736124   
4                           -0.534685                        -0.846595   
...                               ...                              ...   
2942                        -0.376278                        -0.750809   
2943                        -0.320418                        -0.700274   
2944                        -0.118854                        -0.467179   
2945                        -0.205445                        -0.617737   
2946                        -0.072237                        -0.436940   

      angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                        0.006462                              0.162920   
1                       -0.083495                              0.017500   
2                       -0.034956                              0.202302   
3                       -0.017067                              0.154438   
4                       -0.002223                             -0.040046   
...                           ...                                   ...   
2942                    -0.337422            

In [414]:
y_train.value_counts()

Activity
LAYING                1407
STANDING              1374
SITTING               1286
WALKING               1226
WALKING_UPSTAIRS      1073
WALKING_DOWNSTAIRS     986
Name: count, dtype: int64

Since the ACTIVITY column is a categorical column, we need to convert it to numerical encodings known as Label Encoding 

In [415]:
from sklearn.preprocessing import LabelEncoder 

le = LabelEncoder()
y_train = le.fit_transform(y_train)
y_train

array([2, 2, 2, ..., 5, 5, 5])

In [416]:
y_test = test.iloc[:,-1]
y_test

0               STANDING
1               STANDING
2               STANDING
3               STANDING
4               STANDING
              ...       
2942    WALKING_UPSTAIRS
2943    WALKING_UPSTAIRS
2944    WALKING_UPSTAIRS
2945    WALKING_UPSTAIRS
2946    WALKING_UPSTAIRS
Name: Activity, Length: 2947, dtype: object

In [417]:
y_test = le.transform(y_test)
y_test

array([2, 2, 2, ..., 5, 5, 5])

### Here, we are training a logistic regression model before performing the Feature selection

In [418]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lr = LogisticRegression(max_iter=1000)
lr.fit(x_train,y_train)
y_pred = lr.predict(x_test)

In [419]:
accuracy = accuracy_score(y_test,y_pred)
accuracy

0.9596199524940617

### Now we will perform the FFEATURE SELECTION of relevant independent features. 

In [420]:
def remove_duplicate_columns(df):
    
    duplicate_columns = {}
    seen_columns = {}

    for column in df.columns:
        current_column = df[column]

        # Convert column data to bytes
        try:
            current_column_hash = current_column.values.tobytes()
        except AttributeError:
            current_column_hash = current_column.to_string().encode()

        if current_column_hash in seen_columns:
            if seen_columns[current_column_hash] in duplicate_columns:
                duplicate_columns[seen_columns[current_column_hash]].append(column)
            else:
                duplicate_columns[seen_columns[current_column_hash]] = [column]
        else:
            seen_columns[current_column_hash] = column

    return duplicate_columns

### Why Use Hashing?
1. Instead of comparing columns element-wise, we hash the column data for efficient lookup.

2. Faster than comparing every pair of columns manually (which would take O(N^2) time).

In [421]:
# a = pd.DataFrame([10,20,30])
# try:
#     a_bytes = a.values.tobytes()
# except AttributeError:
#     print("AttributeError")
# a_bytes

In [422]:
duplicates = remove_duplicate_columns(x_train)
print(duplicates)

{'tBodyAccMag-mean()': ['tBodyAccMag-sma()', 'tGravityAccMag-mean()', 'tGravityAccMag-sma()'], 'tBodyAccMag-std()': ['tGravityAccMag-std()'], 'tBodyAccMag-mad()': ['tGravityAccMag-mad()'], 'tBodyAccMag-max()': ['tGravityAccMag-max()'], 'tBodyAccMag-min()': ['tGravityAccMag-min()'], 'tBodyAccMag-energy()': ['tGravityAccMag-energy()'], 'tBodyAccMag-iqr()': ['tGravityAccMag-iqr()'], 'tBodyAccMag-entropy()': ['tGravityAccMag-entropy()'], 'tBodyAccMag-arCoeff()1': ['tGravityAccMag-arCoeff()1'], 'tBodyAccMag-arCoeff()2': ['tGravityAccMag-arCoeff()2'], 'tBodyAccMag-arCoeff()3': ['tGravityAccMag-arCoeff()3'], 'tBodyAccMag-arCoeff()4': ['tGravityAccMag-arCoeff()4'], 'tBodyAccJerkMag-mean()': ['tBodyAccJerkMag-sma()'], 'tBodyGyroMag-mean()': ['tBodyGyroMag-sma()'], 'tBodyGyroJerkMag-mean()': ['tBodyGyroJerkMag-sma()'], 'fBodyAccMag-mean()': ['fBodyAccMag-sma()'], 'fBodyBodyAccJerkMag-mean()': ['fBodyBodyAccJerkMag-sma()'], 'fBodyBodyGyroMag-mean()': ['fBodyBodyGyroMag-sma()'], 'fBodyBodyGyroJerk

In [423]:
for i in duplicates:
    print(i,":",duplicates[i])

tBodyAccMag-mean() : ['tBodyAccMag-sma()', 'tGravityAccMag-mean()', 'tGravityAccMag-sma()']
tBodyAccMag-std() : ['tGravityAccMag-std()']
tBodyAccMag-mad() : ['tGravityAccMag-mad()']
tBodyAccMag-max() : ['tGravityAccMag-max()']
tBodyAccMag-min() : ['tGravityAccMag-min()']
tBodyAccMag-energy() : ['tGravityAccMag-energy()']
tBodyAccMag-iqr() : ['tGravityAccMag-iqr()']
tBodyAccMag-entropy() : ['tGravityAccMag-entropy()']
tBodyAccMag-arCoeff()1 : ['tGravityAccMag-arCoeff()1']
tBodyAccMag-arCoeff()2 : ['tGravityAccMag-arCoeff()2']
tBodyAccMag-arCoeff()3 : ['tGravityAccMag-arCoeff()3']
tBodyAccMag-arCoeff()4 : ['tGravityAccMag-arCoeff()4']
tBodyAccJerkMag-mean() : ['tBodyAccJerkMag-sma()']
tBodyGyroMag-mean() : ['tBodyGyroMag-sma()']
tBodyGyroJerkMag-mean() : ['tBodyGyroJerkMag-sma()']
fBodyAccMag-mean() : ['fBodyAccMag-sma()']
fBodyBodyAccJerkMag-mean() : ['fBodyBodyAccJerkMag-sma()']
fBodyBodyGyroMag-mean() : ['fBodyBodyGyroMag-sma()']
fBodyBodyGyroJerkMag-mean() : ['fBodyBodyGyroJerkMag-sm

In [424]:
for one_list in duplicates.values():
    x_train.drop(columns=one_list,inplace=True)
    x_test.drop(columns=one_list,inplace=True)

In [425]:
print(x_train.shape)
print(x_test.shape)

(7352, 541)
(2947, 541)


### We can see that there were multiple duplicate columns, which has been removed and hence the dimensionality of our data has already reduced to 541 from 562

#### Now as the first step, we must perform Variance Threshold, ie. we must drop those column who are having variance less than the Threshold variance.

In [426]:
from sklearn.feature_selection import VarianceThreshold
vt = VarianceThreshold(threshold=0.05)
vt.fit(x_train)

VarianceThreshold(threshold=0.05)

In [427]:
vt.get_support()
# The below values states which columns are have a variance greater than 0.05(represented as True)

array([False, False, False,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True, False,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True, False, False, False,  True, False,  True,
        True,  True,  True,  True,  True,  True,  True, False, False,
       False, False, False, False,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False, False, False,  True,
        True,  True, False, False, False, False,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True, False,  True, False,
        True,  True,  True,  True,  True,  True,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True,  True,
        True,  True,  True, False, False, False,  True,  True,  True,
        True,  True,

In [428]:
sum(vt.get_support())
# There are 350 columns which have a variance > 0.05

350

In [429]:
type(vt.get_support())
# The above dataframes have been converted into numpy arrays.

numpy.ndarray

In [430]:
# It gives you those columns which have a variance greater than 0.05 and drops rest of the columns.
columns = x_train.columns[vt.get_support()]
columns

Index(['tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X', 'tBodyAcc-max()-Y', 'tBodyAcc-max()-Z',
       'tBodyAcc-min()-X',
       ...
       'fBodyBodyGyroJerkMag-skewness()', 'fBodyBodyGyroJerkMag-kurtosis()',
       'angle(tBodyAccMean,gravity)', 'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'subject'],
      dtype='object', length=350)

In [431]:
x_train = vt.transform(x_train)

In [432]:
type(x_train)

numpy.ndarray

In [433]:
x_train = pd.DataFrame(x_train,columns=columns)
x_train.columns

Index(['tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X', 'tBodyAcc-max()-Y', 'tBodyAcc-max()-Z',
       'tBodyAcc-min()-X',
       ...
       'fBodyBodyGyroJerkMag-skewness()', 'fBodyBodyGyroJerkMag-kurtosis()',
       'angle(tBodyAccMean,gravity)', 'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'subject'],
      dtype='object', length=350)

In [434]:
x_train.shape

(7352, 350)

In [435]:
x_test = vt.transform(x_test)
x_test = pd.DataFrame(x_test,columns= columns)
x_test.shape

(2947, 350)

#### Now, after this step where we have dropped some columns just because they had a low variance which could be irrelevant while predicting the dependent variable.
#### Now, we are left with 350 columns instead of 541 columns


In [436]:
correlation = x_train.corr()
correlation

tBodyAcc-std()-X  tBodyAcc-std()-Y  \
tBodyAcc-std()-X                              1.000000          0.927461   
tBodyAcc-std()-Y                              0.927461          1.000000   
tBodyAcc-std()-Z                              0.851668          0.895510   
tBodyAcc-mad()-X                              0.998632          0.922803   
tBodyAcc-mad()-Y                              0.920888          0.997347   
...                                                ...               ...   
angle(tBodyGyroJerkMean,gravityMean)         -0.024810         -0.014592   
angle(X,gravityMean)                         -0.371653         -0.380531   
angle(Y,gravityMean)                          0.471065          0.523600   
angle(Z,gravityMean)                          0.394825          0.433169   
subject                                      -0.064345         -0.115524   

                                      tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
tBodyAcc-std()-X                              0.851668          0.998632   
tBodyAcc-std()-Y                              0.895510          0.922803   
tBodyAcc-std()-Z                              1.000000          0.844469   
tBodyAcc-mad()-X                              0.844469          1.000000   
tBodyAcc-mad()-Y                              0.891441          0.916106   
...                                                ...               ...   
angle(tBodyGyroJerkMean,gravityMean)         -0.006471         -0.024951   
angle(X,gravityMean)                         -0.345011         -0.368191   
angle(Y,gravityMean)                          0.476006          0.466424   
angle(Z,gravityMean)                          0.482828          0.390922   
subject                                      -0.050123         -0.063440   

                                      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  \
tBodyAcc-std()-X                              0.920888          0.846392   
tBodyAcc-std()-Y                              0.997347          0.894509   
tBodyAcc-std()-Z                              0.891441          0.997418   
tBodyAcc-mad()-X                              0.916106          0.839267   
tBodyAcc-mad()-Y                              1.000000          0.891178   
...                                                ...               ...   
angle(tBodyGyroJerkMean,gravityMean)         -0.012341         -0.007231   
angle(X,gravityMean)                         -0.377025         -0.347389   
angle(Y,gravityMean)                          0.525081          0.477607   
angle(Z,gravityMean)                          0.431459          0.479751   
subject                                      -0.114753         -0.055457   

                                      tBodyAcc-max()-X  tBodyAcc-max()-Y  \
tBodyAcc-std()-X                              0.980844          0.895217   
tBodyAcc-std()-Y                              0.917366          0.953573   
tBodyAcc-std()-Z                              0.853884          0.866820   
tBodyAcc-mad()-X                              0.973216          0.889934   
tBodyAcc-mad()-Y                              0.910411          0.949550   
...                                                ...               ...   
angle(tBodyGyroJerkMean,gravityMean)         -0.028871         -0.000297   
angle(X,gravityMean)                         -0.384192         -0.372172   
angle(Y,gravityMean)                          0.480229          0.490324   
angle(Z,gravityMean)                          0.405023          0.427291   
subject                                      -0.055633         -0.095483   

                                      tBodyAcc-max()-Z  tBodyAcc-min()-X  ...  \
tBodyAcc-std()-X                              0.844993         -0.966500  ...   
tBodyAcc-std()-Y                              0.884490         -0.937918  ...   
tBodyAcc-std()-Z                              0.937802         -0.860691  ...   
tBodyAcc-mad()-X                              0.838920        

In [437]:
# Here, we have to drop the columns which have a correlation greater than 0.95

col = correlation.columns
columns_to_drop = []
for i in range(len(col)):
    for j in range(i+1,len(col)):
        if correlation.loc[col[i],col[j]] > 0.95:
            columns_to_drop.append(col[j])

print(len(columns_to_drop))
# some correlation has been counted multiple times, so have have used sets in the next step to remove the duplicates.

1513


In [438]:
columns_to_drop = set(columns_to_drop)
len(columns_to_drop)

197

In [439]:
x_train.drop(columns = columns_to_drop,inplace = True, axis =1)
x_train

tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-max()-Z  \
0            -0.995279         -0.983111         -0.913526         -0.744413   
1            -0.998245         -0.975300         -0.960322         -0.818409   
2            -0.995380         -0.967187         -0.978944         -0.818409   
3            -0.996091         -0.983403         -0.990675         -0.829711   
4            -0.998139         -0.980817         -0.990482         -0.824705   
...                ...               ...               ...               ...   
7347         -0.195387          0.039905          0.077078         -0.076700   
7348         -0.235309          0.004816          0.059280         -0.031966   
7349         -0.218218         -0.103822          0.274533          0.055411   
7350         -0.219139         -0.111412          0.268893          0.055411   
7351         -0.269270         -0.087212          0.177404         -0.133749   

      tBodyAcc-min()-X  tBodyAcc-min()-Y  tBodyAcc-min()-Z  \
0             0.852947          0.685845          0.814263   
1             0.849308          0.685845          0.822637   
2             0.843609          0.682401          0.839344   
3             0.843609          0.682401          0.837869   
4             0.849095          0.683250          0.837869   
...                ...               ...               ...   
7347          0.273480          0.186626          0.266917   
7348          0.282174          0.202880          0.266917   
7349          0.263175          0.202880          0.264670   
7350          0.234850          0.237784          0.264670   
7351          0.234850          0.237784          0.301579   

      tBodyAcc-entropy()-X  tBodyAcc-entropy()-Y  tBodyAcc-entropy()-Z  ...  \
0                -0.407747             -0.679338             -0.602122  ...   
1                -0.714892             -0.500930             -0.570979  ...   
2                -0.592235             -0.485821             -0.570979  ...   
3                -0.627446             -0.850930             -0.911872  ...   
4                -0.786553             -0.559477             -0.761434  ...   
...                    ...                   ...                   ...  ...   
7347              0.195643              0.140158             -0.082077  ...   
7348              0.178138              0.293566             -0.051532  ...   
7349              0.173131              0.276040              0.152003  ...   
7350              0.160333              0.246561              0.021678  ...   
7351              0.312180              0.333536             -0.099101  ...   

      fBodyBodyGyroJerkMag-meanFreq()  fBodyBodyGyroJerkMag-skewness()  \
0                           -0.074323                        -0.298676   
1                            0.158075                        -0.595051   
2                            0.414503                        -0.390748   
3                            0.404573                        -0.117290   
4                            0.087753                        -0.351471   
...                               ...                              ...   
7347                        -0.070157                        -0.588433   
7348                         0.165259                        -0.390738   
7349                         0.195034                         0.025145   
7350                         0.013865                         0.063907   
7351                        -0.058402                        -0.387052   

      angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                       -0.112754                              0.030400   
1                        0.053477                             -0.007435   
2                       -0.118559                              0.177899   
3                       -0.036788                             -0.012892   
4                        0.123320                              0.122542   
...         

In [440]:
x_test.drop(columns = columns_to_drop, inplace = True, axis =1)
x_test.shape

(2947, 153)

#### Now, after removing the highly correlated data we are left with 153 columns and have dropped 197 columns 

In [441]:
from sklearn.feature_selection import f_classif
from sklearn.feature_selection import SelectKBest

sel = SelectKBest(score_func=f_classif,k=100).fit(x_train,y_train)

x_train.columns[sel.get_support()]

Index(['tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-max()-Z', 'tBodyAcc-min()-X', 'tBodyAcc-min()-Y',
       'tBodyAcc-min()-Z', 'tBodyAcc-entropy()-X', 'tBodyAcc-entropy()-Y',
       'tBodyAcc-entropy()-Z', 'tBodyAcc-arCoeff()-X,1',
       'tBodyAcc-arCoeff()-X,2', 'tBodyAcc-arCoeff()-X,3',
       'tBodyAcc-arCoeff()-Y,1', 'tBodyAcc-arCoeff()-Z,1',
       'tBodyAcc-correlation()-X,Y', 'tBodyAcc-correlation()-Y,Z',
       'tGravityAcc-mean()-X', 'tGravityAcc-mean()-Y', 'tGravityAcc-mean()-Z',
       'tGravityAcc-sma()', 'tGravityAcc-energy()-Y', 'tGravityAcc-energy()-Z',
       'tGravityAcc-entropy()-X', 'tGravityAcc-entropy()-Y',
       'tGravityAcc-arCoeff()-Y,1', 'tGravityAcc-arCoeff()-Y,2',
       'tGravityAcc-arCoeff()-Z,1', 'tGravityAcc-arCoeff()-Z,2',
       'tGravityAcc-correlation()-Y,Z', 'tBodyAccJerk-std()-Z',
       'tBodyAccJerk-min()-X', 'tBodyAccJerk-min()-Y', 'tBodyAccJerk-min()-Z',
       'tBodyAccJerk-entropy()-X', 'tBodyAccJerk-arCoeff

In [442]:
columns = x_train.columns[sel.get_support()]
x_train = sel.transform(x_train)
x_train = pd.DataFrame(x_train,columns=columns)

In [443]:
x_train.shape

(7352, 100)

In [444]:
x_test = sel.transform(x_test)
x_test = pd.DataFrame(x_test,columns=columns)


In [447]:
x_test.shape
x_test = pd.DataFrame(x_test,columns = columns)

In [448]:
x_test.shape

(2947, 100)

In [449]:
log_reg = LogisticRegression(max_iter=1000)
log_reg.fit(x_train,y_train)

y_pred = log_reg.predict(x_test)

accuracy = accuracy_score(y_test,y_pred)
accuracy

0.9311163895486936

### Even after removing 460 columns, and performing logistic regression over only 100 columns, the accuracy has been reduced by just 1.07%.
### we can say, The feature selection has worked effectively.